README
NAME
      darkbloom

SYNOPSIS
      darkbloom [--version]

      running 'darkbloom' will start the game
      the '--version' option will cause it to print out version
      information (game version, SDL/SDL_mixer version compiled/linked
      with) and exit

      menu navigation is accomplished with either the mouse or keyboard.
      the up/down arrow keys and/or mouse cursor can be used to select
      menu items. space/enter and/or mouse-clicks are used to activate
      a selected menu item. menu options are changed either with the
      left/right arrow keys or left/right mouse-clicks. escape or right
      click (not on a menu item/option) is used to return to the
      previous menu (exiting the game from the topmost menu).

      the world is navigable using a combination of the keyboard and
      mouse. the arrow keys can be used to attempt to move one space
      in the appropriate direction. alternately the mouse can be used
      to specify a space to walk/run to. the player runs by default,
      however, left clicking with the mouse will cause the player to
      move at walking speed; right clicking will switch to running
      speed. objects are activated by standing next to them, facing
      the appropriate direction, and pressing the enter key. pressing
      the space key will start a boss battle.

      dialogs can be advanced with either mouse clicks or the space,
      enter, or escape keys. dialog choices are selected with either
      the mouse cursor or the up/down arrow keys and are activated by
      either clicking them or pressing either the enter or space key.

      the state of the art battles take place in semi-real-time. you
      select which character to attack with by using the left/right
      arrow keys. the up/down arrow keys are used to select whether to
      execute a melee or magical attack. selecting magic will open
      the spell selection menu; use the up/down arrow keys to select
      which spell to cast. space will return to the previous menu and
      enter will choose the selected attack/spell. once an attack is
      chosen, you use the left/right arrow keys to select which enemy
      to attack. pressing space will return to the menu and pressing
      enter will perform the attack.

OPTIONS
      fullscreen - specifies whether or not to run fullscreen
      sound - whether or not to play sound effects
      music - whether or not to play the background music

      settings are stored to and loaded from the file 'config' in
      the program directory.

FILE FORMATS BELOW

ENEMIES
      name: ENEMY_NAME
      hp: HITPOINTS
      strength: STRENGTH
      agility: AGILITY
      xp: EXPERIENCE_POINTS_IT_IS_WORTH
      graphic_idle: IDLE_GRAPHIC_FILENAME
      graphic_attack: ATTACK_GRAPHIC_FILENAME
      offset_x: X_DRAWING_OFFSET
      offset_y: Y_DRAWING_OFFSET

      there is no difference between an enemy datafile and a boss
      datafile other than the extension usually given to them.

      example:
        name: Rat
        hp: 20
        strength: 12
        agility: 22
        xp: 20
        graphic_idle: enemies/images/rat_idle.png
        graphic_attack: enemies/images/rat_attack.png
        offset_x: 0
        offset_y: 50

ENEMYLISTS
      enemy definition filename
      
      example:
        enemies/rat.enemy
        enemies/your_mom.enemy

OBJECTS
      name: OBJECT_NAME
        string (first 10 chars are the only ones displayed)
        should not contain spaces, i think.
      visible: IS_IT_VISIBLE
        1 or 0
      persistent: SHOULD_WE_RELOAD_IT
        1 or 0
        if the object is persistent, it will not be reloaded when we
        return to the objects area. this way it will retain any
        changes that have been made to it. only 1 object of any
        given name can be persistent in an area (more or less)
      active: IS_IT_CURRENTLY_ACTIVE
        1 or 0
        whether the object should be placed in the world effectively
        on load, or just loaded
      zindex: VISIBLE_LAYER_ORDERING
        a float
        this is used to determine which order objects are drawn. an
        object of zindex 1 will always be below (visibly) an object of
        zindez 2.
      offset_x: X_DRAWING_OFFSET
        signed integer
        how far (on the x axis) to offset from the expected position
        when drawing the object
      offset_y: Y_DRAWING_OFFSET
        signed integer
        how far (on the y axis) to offset from the expected position
        when drawing the object
      image: GRAPHIC_FILENAME
        NONE or filename (80 chars max)
      frame_size.x: FRAME_X_FOR_GRAPHIC
      frame_size.y: FRAME_Y_FOR_GRAPHIC
      frame_size.w: FRAME_W_FOR_GRAPHIC
      frame_size.h: FRAME_H_FOR_GRAPHIC
        integers
        objects have 'facings'. that is, they can point in different
        directions and the frame size is how you specify how big
        the actual object's image is (since the graphic may contain
        several facings)
      a_script: ACTIVATION_SCRIPT_FILENAME
        NONE or filename (80 chars max)
        the script to run if this object is activated
      c_script: COLLISION_SCRIPT_FILENAME
        NONE or filename (80 chars max)
        the script to run if this object is collided with
      collidable: COLLIDABILITY TYPE
        NONE or FULL or MOVE
        NONE = only triggers the collision script
        FULL = immovable
        MOVE = pushable object. if there is not another object in the
        way, the object may be pushed into a new space.

      example:
        name: Tree
        visible: 1
        persistent: 0
        active: 1
        zindex: 2
        offset_x: -32
        offset_y: -64
        image: images/tree.png
        frame_size.x: 0
        frame_size.y: 0
        frame_size.w: 96
        frame_size.h: 96
        a_script: scripts/tree.script
        c_script: NONE
        collidable: FULL

ANIMATED OBJECTS
      same as OBJECTS with the 'image: ' line replaced by:
      image_down: DOWN_FACING_ANIMATION
        NONE or filename
      move_frame_start: INTEGER
        integer
        which frame the movement animation starts on (if any)
      and three more for up, left, and right facings.

      example:
        name: Yddor
        visible: 1
        persistent: 1
        active: 1
        zindex: 2
        offset_x: 0
        offset_y: -32
        image_down: images/rod_anim_down.png
        move_frame_start: 2
        image_up: images/rod_anim_up.png
        move_frame_start: 2
        image_left: images/rod_anim_left.png
        move_frame_start: 2
        image_right: images/rod_anim_right.png
        move_frame_start: 2
        frame_size.x: 0
        frame_size.y: 0
        frame_size.w: 32
        frame_size.h: 64
        a_script: NONE
        c_script: NONE
        collidable: FULL

DIALOGS
      $(VARNAME) string your mom
        creates or sets a variable named VARNAME of type 'string' with
        the value 'your mom'. valid types are int/float/bool/string.

      [1]
        denotes a section (in this case, section 1). used by jumps.
        the closing bracket is not required.

      @1
        jumps to the section number specified. target section number
        can be a variable (@ $(EXIT_SECTION) for instance)

      !$(VARNAME)
        evaluates the variable (in this case VARNAME) as true or false
        and skips the next command if the eval was false

      dialog system special variables
      the effect will take place on the next dialogbox displayed
      
      DBOXTOP
        evaluated as: bool
        effect:
          if true dialogbox is drawn at the top of the screen
          else it is drawn at the bottom
      
      DBOXTITLE
        evaluated as: string
        effect:
          used to set the title/caption of the dialogbox
      
      DBOXIMG
        evaluated as: string
        effect:
          determines the image file to load and display in the
          dialogbox
      
      DBOXSND
        evaluated as: string
        effect:
          determines the sound file to load and play
          the value is auto automatically cleared and the sound stopped
          when the dialogbox is closed
      
      DBOXSNDLOOPS
        evaluated as: int
        effect:
          determines how many additional times to play the dialogbox
          sound. a value of 1 plays the sound effect 2 times total. -1
          will attempt to loop forever. automatically reset to 0 when
          closed.


SCRIPTING
      Lfilename
        load filename as the current level
      P x y
        specifies new x, y coordinates for player
      Dfilename
        specifies a dialog to execute
        sets DIALOG_RETVAL as the exitcode of the dialog
      d msecs
        delay in milliseconds
      Mfilename
        load filename as the new music track
      Sfilename
        load filename and play as sound effect
        the sound effect chunk is freed when the parent script
        returns, so you may need a delay if you want to make sure
        it plays long enough
      W property = value
        sets the 'who' object property 'property' to the value of 'value'
      w property = value
        sets the 'what' object property 'property' to the value of 'value'
      B num
        starts a boss battle w/ 'num' random henchmen
      b num
        starts a battle w/ 'num' random henchmen
      $, [, @, ! all work the same as in the dialog system

MAPS
      AREANAME
      ENEMYLIST
      TILELIST
      AREA_WIDTH AREA_HEIGHT
      TILE_ARRAY
      NUMBER_OF_OBJECTS
      OBJECT_X OBJECT_Y
      OBJECT_FILENAME

      maps have a minimum width/height of 15 and a maximum of 100.
      objects outside the area bounds will not be drawn.

      example:
        Darkshire
        
        enemies/start.enemylist
        
        tiles/common.tilelist
        
        15 20
        
        010 010 010 010 010 010 010 010 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 010 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 010 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 010 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 010 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 010 010 010 010 010 010 010 010
        010 010 010 010 010 041 041 041 041 041 041 010 010 010 010
        010 010 010 010 010 041 010 010 010 010 041 010 010 010 010
        010 010 010 010 010 041 010 010 010 010 041 010 010 010 010
        010 010 010 010 010 041 010 010 010 010 041 010 010 010 010
        010 010 010 010 010 041 010 010 010 010 041 010 010 010 010
        010 010 010 010 010 041 041 010 041 041 041 010 010 010 010
        010 010 010 010 010 010 010 040 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 040 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 040 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 040 010 010 010 010 010 010 010
        010 010 010 010 010 010 010  40 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 040 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 040 010 010 010 010 010 010 010
        010 010 010 010 010 010 010 040 010 010 010 010 010 010 010

        10

        17 17
        objects/tree.object

        8 3
        objects/battle.object

        19 19
        objects/battle.object

        4 13
        objects/tree.object

        10 14
        objects/deadtree.object

        6 3
        objects/deadtree.object

        7 11
        objects/house_bottom.object

        7 11
        objects/house_top.object

        9 10
        objects/catsailor.anim

        15 5
        objects/swirl.object


TILELISTS
      tilenumber and filename pairs. tilenumber may be preceeded by
      spaces or zeroes.

      example:
        010 tiles/grass.png
        040 tiles/stone.png
        041 tiles/stone.png
         11 tiles/sand.png
        020 tiles/dirt.png
          2 tiles/ltdirt_brgrass.png

BUGS
      mostly harmless

CONTRIBUTORS
      Roderick L. E. Newill
      Paul W. Carley

EOF


CHANGES
darkbloom
----------------------------------------------------------------------

Fri Oct  7 10:51:51 EDT 2005
  -fix: clipping rect still in effect when scripts are called from
    collisions in Object.move()

Tue Aug 30 09:19:45 EDT 2005
  -use 'strrchr' instead of 'rindex' to support mingw

Sun Aug 14 02:01:01 EDT 2005
  -removed: world_graphics directory, old (numbered) maps. replaced
    world_graphics images w/ named images (like grass.png) in tiles
    directory
  -added: Path class (part of base Object class) to handle NPC routes
    currently there is no way to add a path, since i haven't decided
    how exactly they should be specified (relative/absolute, from a
    script or from a file, should they have delays? should they
    execute scripts when you reach the waypoint?)

Fri Aug 12 11:36:52 EDT 2005
  -added: Object::move() function to handle the moving of an object
    along a path. (currently it just moves the object as far as the
    move_stack tells it to, since there is no path information. (only
    the player's mouse navigation currently uses this)

Thu Aug 11 12:11:49 EDT 2005
  -added: segfault caused by attempting to insert animated objects
    into the ObjectLayerList in the middle AND at the end if they were
    supposed to go in the middle.
  -fixed: segfault caused by attempting to insert animated objects
    into the ObjectLayerList in the middle AND at the end if they were
    supposed to go in the middle.
  -maps now load tile graphic information from a tilelist (specified
    as part of the map/level format)
  -animated objects can be specified in the object section of the map
    file format (they must end in .anim however)

Fri Aug  5 13:32:11 EDT 2005
  -remodelled worldloop to support animated object idle animations
    it now runs smoother as well
  -worldloop speed depends on value of 'msecs_per_frame' (global var)
  -fixed: animated object idle animations only using frame 0
  -fixed: animated object movement animation starting on frame 0
  -added: bool flushJunkEvents(SDL_Event *event) (video.(hh/cc)
  -updated: rod_anim_down.(png/xcf) to have an idle animation
  -idle animation frame interval temporarily hardcoded at 250ms until
    per frame intervals are added to the animated object format, and
    the movement animation frame interval to 100ms
  -mouse clicks while walking to a tile, set the new tile as the
    destination (you can change direction before you walk all the way)

Tue Aug  2 15:37:25 EDT 2005
  -nmystuff and bossstuff structs merged into Enemy class
    life changed to hp, magic to mp, str to strength, etc
  -enemylist specification added to map file format
  -enemies now loaded on a per area basis
  -fixed several battle crashes when < 3 enemy types are loaded
  -battle fonts are now freed in the battle destruct function
  -new enemy type graphics: Rat, GiantRat
  -added BUGS file to keep track of known bugs/issues
  -enemy x, y offsets now respected when drawing enemies
  -removed ed's unused code

Fri Jul 29 04:55:25 EDT 2005
  -use CXX instead of CPP in Makefile
  -battle system added/started (roderick)
  -stack added (roderick), converted to use Direction enum
  -lots of .cc & .hh files from ed
  -initial (and probably final) script system
  -interpolateVars() now takes a VAR_paramList to interpolate from
    ditto for processVar()
  -A* pathfinding (rod)
    fix: dyslexic pathfinding (rod)
    fix: walking away when you can't get any closer (pathfinding) (rod)
  -rewrote most of rod's world code (now it's World class code!) to
    work w/ the Object classes better
  -Object and AnimatedObject classes completed
  -proper object layering
  -rewrote 'worldLoop' (the game loop, basically)
  -collisions and activation trigger scripts
  -connected mouseclicks to pathfinding/integrated pathfinding w/
    objects
  -changed a bunch of SDL_GetError calls to more appropriate
    IMG_GetError
  -Object collisions/Object movement
  -animated player moving (from tile to tile)
  -removed rod's world moving and collision functions
  -WHO KNOWS WHAT ELSE, ONLY CVS CAN SAY FOR SURE

Mon Jul 25 02:08:43 EDT 2005
  -added a 'poisoned' visual effect to the sidebar
  -added beginnings of an 'Object' class
  -changed equiped weapon sidebar images to 45x45
  -added what will become a map editor (and its makefile)

Sun Jul 24 18:33:22 EDT 2005
  -sidebar added
  -maps updated to include an area name

Sat Jul 23 18:03:55 EDT 2005
  -integrated rod's world and plyr code
  -added some error checking to map loader/plyr code
  -tiles are expected to be 3 digits (leading zeros accepted)
  -changes to world.hh #defines (WIDTH -> TILEWIDTH etc)
  -added map and world_graphics directories

Sat Jul 23 10:50:36 EDT 2005
  -sound effect WAVs reencoded as .ogg

Sat Jul 23 00:11:46 EDT 2005
  -removed some debugging output
  -updated sounds/music/fullscreen toggles
  -fullscreen workaround for win32 (mouse sometimes leaves artifacts)
  -updateDisplay() replaces all direct calls to SDL_Flip
  -mouse and sound support for dialog boxes
  -version information available as a command-line option
  -changed music to 'midori hirano - i was'
  -updated license information in sounds directory
  -fixed: flicker when displaying dialog boxes in sequence
  -music and sound enabled by default
  -some code reformatting and commenting
  -separated logos into own functions
  -sample outro added
  -allow sounds/music/fullscreen toggle at any point in game
  -removed dialog testapp (it wasn't much anyway)
  -added toInt() function to params and parameter lists
  -updated dev-c++ project
  -don't open the display window until graphics are loaded
  -new sound function to load and begin playing direct from file
  -all sounds stopped and freed manually on exit
  -fixed: segfault on inserting/setting a var w/ NULL value pointer
  -removed all opengl video code (it was already commented out)
  -delay() is escapeable by default

Fri Jul 22 00:22:04 EDT 2005

src/var.cc
/*********************************************************************
variable handling structures

pcarley@lssu.edu
(c) trained circus idiot, ink.
this is GPL'ed software (see COPYING for license)

modified:
  Tue Jul 19 23:04:56 EDT 2005
    -added setData(), toString() and eval() functions to VAR_param
  Tue Jul 19 13:21:16 EDT 2005
    -added flush() and get() functions to VAR_paramList
    -add() renamed to set() and vars are now type and value switched
    -removed some debugging statements
  Thu Jun 23 22:28:05 EDT 2005

*********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "var.hh"

std::string var_error;

const char* VAR_GetError()
{
  return var_error.c_str();
}


void VAR_SetError(const char* error)
{
  var_error = error;
}


/*********************************************************************
VAR_param constructor
  initializes the structure as empty
*********************************************************************/
VAR_param::VAR_param()
{
  key = "";
  type = VAR_STRING;
  data.string_data = NULL;
  persistent = false;
  next = NULL;
}


/*********************************************************************
VAR_param constructor
  initializes the structure with the appropriate keyvalue
*********************************************************************/
VAR_param::VAR_param(const char *k)
{
  key = k;
  type = VAR_STRING;
  data.string_data = NULL;
  persistent = false;
  next = NULL;
}


/*********************************************************************
VAR_param destructor
  frees up the allocated string if any
*********************************************************************/
VAR_param::~VAR_param()
{
  if(type == VAR_STRING && data.string_data != NULL)
  {
    delete data.string_data;
    data.string_data = NULL;
  }
}


/*********************************************************************
VAR_param setData()
  sets the data type and value
*********************************************************************/
void VAR_param::setData(VAR_datatype datatype, const void *value)
{
  /*delete old string, if there was one*/
  if(type==VAR_STRING && data.string_data!=NULL) delete data.string_data;

  type = datatype;
  switch(type)
  {
    case VAR_STRING:
      data.string_data = new std::string;
      if(value == NULL) *(data.string_data) = "";
      else *(data.string_data) = (char*)value;
      break;
    case VAR_BOOL:
      if(value == NULL) data.bool_data = false;
      else data.bool_data = *((bool*)value);
      break;
    case VAR_INT:
      if(value == NULL) data.int_data = 0;
      else data.int_data = *((int*)value);
      break;
    case VAR_FLOAT:
      if(value == NULL) data.float_data = 0;
      else data.float_data = *((float*)value);
      break;
  }
}


/*********************************************************************
VAR_param eval()
  returns the value as true or false
*********************************************************************/
bool VAR_param::eval()
{
  switch(type)
  {
    case VAR_STRING:
      if(data.string_data == NULL) return false;
      if(*(data.string_data)->c_str()) return true;
      break;
    case VAR_BOOL:
      return data.bool_data;
    case VAR_INT:
      return (bool)data.int_data;
    case VAR_FLOAT:
      return (bool)data.float_data;
  }
  return false;
}



/*********************************************************************
VAR_param toString()
  converts the value to a string
*********************************************************************/
char* VAR_param::toString(char *buffer, int num)
{
  switch(type)
  {
    case VAR_STRING:
      if(data.string_data == NULL) return NULL;
      strncpy(buffer, data.string_data->c_str(), num-1);
      buffer[num-1] = '\0';
    break;
    case VAR_BOOL:
      if(data.bool_data) strcpy(buffer, "true");
      else strcpy(buffer, "false");
      break;
    case VAR_INT:
      snprintf(buffer, num, "%i", data.int_data);
      break;
    case VAR_FLOAT:
      snprintf(buffer, num, "%.2f", data.float_data);
      break;
  }
  return buffer;
}


/*********************************************************************
VAR_param toInt()
  converts the value to a string
*********************************************************************/
int VAR_param::toInt()
{
  switch(type)
  {
    case VAR_STRING:
      return 0;
    case VAR_BOOL:
      if(data.bool_data) return 1;
      else return 0;
    case VAR_INT:
      return data.int_data;
    case VAR_FLOAT:
      return (int)data.float_data;
  }
  return 0;
}


/*********************************************************************
VAR_paramList constructor
  initializes the structure as empty
*********************************************************************/
VAR_paramList::VAR_paramList()
{
  head = NULL;
}


/*********************************************************************
VAR_paramList destructor
  steps through all nodes, freeing the used memory
*********************************************************************/
VAR_paramList::~VAR_paramList()
{
  /*local variables*/
  VAR_param *search, *temp;

  search = head;
  while(search != NULL)
  {
    temp = search;
    search = search->next;
    delete temp;
  }
  head = NULL;
}


/*********************************************************************
VAR_paramList set(const char *param,
                  const char *datatype,
                  const void *value);
  returns the VAR_param that was added (or alternately the one that
  was already attached to that paramname) on successful addition of
  the parameter or NULL on invalid type specification or other error
*********************************************************************/
VAR_param* VAR_paramList::set(const char *param,
                              const char *datatype,
                              const void *value)
{
  /*local variables*/
  VAR_param *node, *search;
  VAR_datatype type;
  std::string typestring;

  typestring = datatype;
  /*convert typestring to type enum*/
  if(typestring == "string") type = VAR_STRING;
  else
    if(typestring == "bool") type = VAR_BOOL;
    else
      if(typestring == "int") type = VAR_INT;
      else
        if(typestring == "float") type = VAR_FLOAT;
        else
        {
          fprintf(stderr, "DEBUG: invalid type specification: %s\n",
                  typestring.c_str());
          return NULL;
        }

  search = head;

  /*not an empty list?*/
  if(head != NULL)
  {
    /*find node in front of insertion position or previous instance*/
    while((search->next != NULL) && (search->next->key <= param))
    {
      search = search->next;
    }
    /*already in list?*/
    if(search->key == param)
    {
      search->setData(type, value);
      return search;
    }
  }

  /*create new node*/
  node = new VAR_param(param);
  node->setData(type, value);

  /*insert at front?*/
  if((head == NULL) || (search == head && search->key > param))
  {
    node->next = search;
    head = node;
  } else
    /*insert at end?*/
    if(search->next == NULL)
    {
      search->next = node;
      node->next   = NULL;
    } else
      /*insert in middle*/
      {
        node->next   = search->next;
        search->next = node;
      }

  /*return the added node*/
  return node;
}


/*********************************************************************
VAR_paramList get(const char *param);
  returns the VAR_param that is attached to that paramname or NULL if
  the parameter has not been set.
*********************************************************************/
VAR_param* VAR_paramList::get(const char *param)
{
  /*local variables*/
  VAR_param *search;

  search = head;

  /*not an empty list?*/
  if(search != NULL)
  {
    /*find node*/
    while((search->key != param) && (search->next != NULL))
    {
      search = search->next;
    }
    /*in list?*/
    if(search->key == param) return search;
  }
  /*not in list*/
  return NULL;
}


/*********************************************************************
VAR_paramList perist();
  marks the parameter as persistent (adding it if it doesn't exist)
  returns the VAR_param that was added (or alternately the one that
  was already attached to that paramname)
*********************************************************************/
VAR_param* VAR_paramList::persist(const char *param)
{
  /*local variables*/
  VAR_param *node, *search;

  search = head;

  /*not an empty list?*/
  if(head != NULL)
  {
    /*find node in front of insertion position or previous instance*/
    while((search->next != NULL) && (search->next->key <= param))
    {
      search = search->next;
    }
    /*already in list?*/
    if(search->key == param)
    {
      search->persistent = true;
      return search;
    }
  }

  /*create new node*/
  node = new VAR_param(param);
  node->persistent = true;

  /*insert at front?*/
  if((head == NULL) || (search == head && search->key > param))
  {
    node->next = search;
    head = node;
  } else
    /*insert at end?*/
    if(search->next == NULL)
    {
      search->next = node;
      node->next   = NULL;
    } else
      /*insert in middle*/
      {
        node->next   = search->next;
        search->next = node;
      }

  /*return the added node*/
  return node;
}


/*********************************************************************
VAR_paramList eval(const char *param);
  returns true or false for the specified parameter based on the
  existence and/or value of the parameter
*********************************************************************/
bool VAR_paramList::eval(const char *param)
{
  /*local variables*/
  VAR_param *search;

  search = head;

  /*not an empty list?*/
  if(search != NULL)
  {
    /*find node*/
    while((search->key != param) && (search->next != NULL))
    {
      search = search->next;
    }
    /*in list?, then returns its evaluated value*/
    if(search->key == param) return search->eval();
  }
  /*not in list, so it is false*/
  return false;
}


/*********************************************************************
VAR_paramList toString(const char *param, char *buffer, int num);
  copies the string representation of the variable to the buffer (at
  most num bytes in length) and returns a pointer to the buffer or
  NULL if the variable is not in the list
*********************************************************************/
char* VAR_paramList::toString(const char *param, char *buffer, int num)
{
  /*local variables*/
  VAR_param *search;

  search = head;

  /*not an empty list?*/
  if(search != NULL)
  {
    /*find node*/
    while((search->key != param) && (search->next != NULL))
    {
      search = search->next;
    }
    /*in list?, then returns its evaluated value*/
    if(search->key == param) return search->toString(buffer, num);
  }
  /*not in list, so it is false*/
  return NULL;
}


/*********************************************************************
VAR_paramList toInt(const char *param);
  returns the integer representation of the variable, (0 if it has not
  been set)
*********************************************************************/
int VAR_paramList::toInt(const char *param)
{
  /*local variables*/
  VAR_param *search;

  search = head;

  /*not an empty list?*/
  if(search != NULL)
  {
    /*find node*/
    while((search->key != param) && (search->next != NULL))
    {
      search = search->next;
    }
    /*in list?, then returns its evaluated value*/
    if(search->key == param) return search->toInt();
  }
  /*not in list, so it is 0*/
  return 0;
}


/*********************************************************************
VAR_paramList flush()
  removes all non-persistent values from list
*********************************************************************/
void VAR_paramList::flush()
{
  /*local variables*/
  VAR_param *search, *prev, *temp;

  search = head;
  prev = NULL;
  while(search != NULL)
  {
    /*if it is a keeper*/
    if(search->persistent)
    {
      /*if the isn't the first node, set the previous node's next pointer*/
      if(prev != NULL) prev->next = search;
      /*else set the head pointer to point here*/
      else head = search;
      /*set this as the previous preserved node*/
      prev = search;
      search = search->next;
      continue;
    }
    /*delete this param*/
    temp = search;
    search = search->next;
    delete temp;
  }
  /*make sure list is NULL terminated*/
  if(prev != NULL) prev->next = NULL;
  else head = NULL;
}


/*********************************************************************
VAR_paramList clear()
  removes all values from list
*********************************************************************/
void VAR_paramList::clear()
{
  /*local variables*/
  VAR_param *search, *temp;

  search = head;
  while(search != NULL)
  {
    temp = search;
    search = search->next;
    delete temp;
  }
  head = NULL;
}


/*********************************************************************
VAR_paramList debug()
  prints out debugging information on all nodes
*********************************************************************/
void VAR_paramList::debug()
{
  /*local variables*/
  VAR_param *search;
  int nodes;

  printf("\nparamList: DEBUG\n");

  search = head;
  nodes  = 0;
  while(search != NULL)
  {
    printf("  key: %s\n", search->key.c_str());
    switch(search->type)
    {
      case VAR_STRING:
        printf("value: %s\n", search->data.string_data->c_str());
      break;
      case VAR_BOOL:
        if(search->data.bool_data) printf("value: true\n");
        else printf("value: false\n");
        break;
      case VAR_INT:
        printf("value: %i\n", search->data.int_data);
        break;
      case VAR_FLOAT:
        printf("value: %f\n", search->data.float_data);
        break;
    }
    search = search->next;
    nodes++;
  }

  printf("paramList: %i nodes total\n\n", nodes);
}

/*EOF*/
